framework/src/relationshipmanagers/RedistributeProperties.C

```diff
@@ -0,0 +1,209 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "RedistributeProperties.h"
+
+#include "MaterialProperty.h"
+#include "MaterialPropertyStorage.h"
+
+#include <timpi/parallel_sync.h>
+
+registerMooseObject("MooseApp", RedistributeProperties);
+
+InputParameters
+RedistributeProperties::validParams()
+{
+  InputParameters params = RelationshipManager::validParams();
+
+  return params;
+}
+
+RedistributeProperties::RedistributeProperties(const InputParameters & parameters)
+  : RelationshipManager(parameters)
+{
+}
+
+void
+RedistributeProperties::operator()(const MeshBase::const_element_iterator &,
+                                   const MeshBase::const_element_iterator &,
+                                   processor_id_type,
+                                   map_type &)
+{
+}
+
+std::unique_ptr<GhostingFunctor>
+RedistributeProperties::clone() const
+{
+  return std::make_unique<RedistributeProperties>(*this);
+}
+
+std::string
+RedistributeProperties::getInfo() const
+{
+  return "RedistributeProperties";
+}
+
+// the LHS ("this" object) in MooseApp::addRelationshipManager is the existing RelationshipManager
+// object to which we are comparing the rhs to determine whether it should get added
+bool
+RedistributeProperties::operator>=(const RelationshipManager & rhs) const
+{
+  const auto * rm = dynamic_cast<const RedistributeProperties *>(&rhs);
+
+  // All RedistributeProperties objects are effectively equivalent
+  return rm;
+}
+
+void
+RedistributeProperties::addMaterialPropertyStorage(
+    std::vector<std::shared_ptr<MaterialData>> & mat_data, MaterialPropertyStorage & mat_props)
+{
+  _materials.emplace_back(&mat_data, &mat_props);
+}
+
+void
+RedistributeProperties::redistribute()
+{
+  const MeshBase & mesh = _moose_mesh->getMesh();
+  const processor_id_type pid = mesh.processor_id();
+
+  for (auto & [mat_data, mat_prop_store] : _materials)
+  {
+    // Once we've redistributed data elsewhere we'll delete it here.
+    //
+    // We don't need stored properties on non-local elements, and so
+    // we don't compute them and we don't *bother deleting them when
+    // non-local elements are coarsened away*, so this may be our last
+    // chance to avoid dangling pointers in future redistribute()
+    // calls.
+    std::set<const Elem *> elems_being_migrated;
+
+    if (mat_prop_store->hasStatefulProperties())
+    {
+      MaterialData & my_mat_data = *((*mat_data)[/*_tid*/ 0]); // Not threaded
```

Comment:
Can we assert then that we are `!in_threads`? I get nervous after my coupling functor was not thread safe (I know that you are only adding this to the mesh but still like to assert)

Reply:
Kinda implied by the API that's being called, IMHO (any threading ought to be done inside redistribute() implementations, and the fact that it's a function applied to a whole mesh with no range-of-elements should be enough to deduce that), but I can't actually bring myself to say the phrase "this code shouldn't have so many assertions", so I'll add it.